lib/spack/spack/new_installer.py

```diff
@@ -0,0 +1,1194 @@
+        elif not restage:
+            raise NotImplementedError("Restaging builds is not implemented")
+        elif skip_patch:
+            raise NotImplementedError("Skipping patches is not implemented")
+        elif stop_at is not None:
+            raise NotImplementedError("Stopping at an install phase is not implemented")
+        elif stop_before is not None:
+            raise NotImplementedError("Stopping before an install phase is not implemented")
+        elif tests is not False:
+            raise NotImplementedError("Tests during install are not implemented")
+        elif use_cache:
+            raise NotImplementedError("Using cache only is not implemented")
+        # verbose and concurrent_packages are not worth erroring out for
+
+        specs = [pkg.spec for pkg in packages]
+
+        # Buffer for incoming, partially received state data from child processes
+        self.state_buffers: Dict[int, str] = {}
+
+        #: lookup package by unique identifier
+        self.nodes = {spec.dag_hash(): spec for spec in spack.traverse.traverse_nodes(specs)}
+
+        #: mapping from parent to children (children are deleted installed)
+        self.parent_to_child = {
+            parent.dag_hash(): {child.dag_hash() for child in parent.dependencies()}
+            for parent in self.nodes.values()
+        }
+
+        #: reverse mapping from child to parents
+        self.child_to_parent: Dict[str, Set[str]] = {}
+        for parent, children in self.parent_to_child.items():
+            for child in children:
+                if child not in self.child_to_parent:
+                    self.child_to_parent[child] = set()
+                self.child_to_parent[child].add(parent)
+
+        # 1. Assign an install prefix to each spec
+        # 2. Prune the build graph of already installed packages
+        db = spack.store.STORE.db
+        with db.read_transaction():
+            for key in list(self.nodes):
+                spec = self.nodes[key]
+                _, record = db.query_by_spec_hash(key)
+                if record and record.path:
+                    spec.set_prefix(record.path)
+                else:
+                    spec.set_prefix(spack.store.STORE.layout.path_for_spec(spec))
+                if not record:
+                    continue
+                # Remove node: wire children to parents and vice versa, delete node from graph.
+                for parent in self.child_to_parent.get(key, ()):
+                    self.parent_to_child[parent].remove(key)
+                    self.parent_to_child[parent].update(self.parent_to_child.get(key, ()))
+                for child in self.parent_to_child.get(key, ()):
+                    self.child_to_parent[child].remove(key)
+                    self.child_to_parent[child].update(self.child_to_parent.get(key, ()))
+                self.parent_to_child.pop(key, None)
+                self.child_to_parent.pop(key, None)
+                del self.nodes[key]
```

Comment:
It looks like this is assuming that a db record only exists if the package is installed, but that's not a safe assumption. We need to explicitly check `record.installed` to know whether it's installed (it could be a deleted package depended on by something else, for example).

Reply:
Good catch, that was indeed a bug. I've incorporated this in a bigger initialization change: build deps are now only conditionally enqueued like in installer.py.